symengine/uint_base.h

```diff
@@ -0,0 +1,106 @@
+/**
+ *  \file polynomial_int.h
+ *  Class for Univariate Polynomial Base
+ **/
+#ifndef SYMENGINE_UINT_BASE_H
+#define SYMENGINE_UINT_BASE_H
+
+#include <symengine/basic.h>
+
+namespace SymEngine
+{
+
+template <typename Container, typename Poly>
+class UIntPolyBase : public Basic
+{
+public:
+    RCP<const Symbol> var_;
+    Container poly_;
+
+    UIntPolyBase(const RCP<const Symbol> &var, Container &&container)
+        : var_{var}, poly_{container}
+    {
+    }
+
+    // unify these two constructor? another template would be required
+    // may solve some more problems, like `get_degree` virtualization
+    UIntPolyBase(const RCP<const Symbol> &var,
+                 const std::vector<integer_class> &v)
+        : var_{var}
+    {
+    }
+
+    UIntPolyBase(const RCP<const Symbol> &var, const std::vector<Expression> &v)
+        : var_{var}
+    {
+    }
+
+    // TODO think of something to make this purely virtual
+    //! \returns the degree of the polynomial
+    // virtual unsigned int get_degree() const = 0;
+
+    //! \returns `-1`,`0` or `1` after comparing
+    virtual int compare(const Basic &o) const = 0;
+
+    //! \returns `true` if two objects are equal
+    inline bool __eq__(const Basic &o) const
+    {
+        return eq(*var_, *(static_cast<const Poly &>(o).var_))
+               and poly_ == static_cast<const Poly &>(o).poly_;
+    }
+
+    inline RCP<const Symbol> get_var() const
+    {
+        return var_;
+    }
+
+    inline const Container &get_poly() const
+    {
+        return poly_;
+    }
+
+    friend RCP<const Poly> add_poly(const Poly &a, const Poly &b)
+    {
+        RCP<const Symbol> var = symbol("");
+        if (a.var_->get_name() == "") {
+            var = b.var_;
+        } else if (b.var_->get_name() == "") {
+            var = a.var_;
+        } else if (!(a.var_->__eq__(*b.var_))) {
+            throw std::runtime_error("Error: variables must agree.");
+        } else {
+            var = a.var_;
+        }
```

Comment:
What should happen in the case that the coefficients don't match is to return a multivariate polynomial even if one is univariate and the other is a coefficient. 
```
Poly(x, x) + Poly(2, y)                                                                                                 
Out[6]: Poly(x + 2, x, y, domain='ZZ') 
```
Remember, these kinds of design decisions are easy because SymPy has them and we should follow SymPy whenever possible except when it may affect performance.